### PR TITLE
Changes to nuget functionality

### PR DESCRIPTION
**Pull request by freddydk**
With this release the implementation for delivering to NuGet packages (by adding the NuGetContext secret), is similar to the functionality behind delivering to GitHub packages and the implementation is final.

New Settings:
- `trustMicrosoftNuGetFeeds` Unless this setting is set to false, AL-Go for GitHub will trust the NuGet feeds provided by Microsoft. The feeds provided by Microsoft contains all Microsoft apps, all Microsoft symbols and symbols for all AppSource apps.
- `trustedNuGetFeeds` - can be an array of NuGet feed specifications, which AL-Go for GitHub will use for dependency resolution. Every feed specification must include a url property and can optionally include a few other properties:
  - Url - The URL of the feed (examples: https://pkgs.dev.azure.com/myorg/apps/\_packaging/myrepo/nuget/v3/index.json or https://nuget.pkg.github.com/mygithuborg/index.json")
  - Patterns - AL-Go for Github will only trust packages, where the ID matches this pattern. Default is all packages (\*).
  - FingerPrints - If specified, AL-Go for GitHub will only trust packages signed with one of the fingerprints specified in this array of fingerprints.
  - AuthTokenSecret - If the NuGet feed specified by URL is private, the authTokenSecret must be the name of a secret containing the authentication token with permissions to search and read packages from the NuGet feed

### CODE REVIEW

**Comment by lippertmarkus** [comment]
eagerly waiting for this being merged so our product development can start using AL-Go

File: RELEASENOTES.md [hunk 1]
```diff
@@ -2,6 +2,20 @@
 
 - Issue 1184 Publish to Environment fails on 'Permission Denied'
 
+### New Settings
+
+- `deliverTo<deliverytarget>` now has an additional property called `ContinuousDelivery`, indicating whether or not to run continuous delivery to this deliveryTarget. Default is true.
+- `trustMicrosoftNuGetFeeds` Unless this setting is set to false, AL-Go for GitHub will trust the NuGet feeds provided by Microsoft. The feeds provided by Microsoft contains all Microsoft apps, all Microsoft symbols and symbols for all AppSource apps.
+- `trustedNuGetFeeds` - can be an array of NuGet feed specifications, which AL-Go for GitHub will use for dependency resolution. Every feed specification must include a url property and can optionally include a few other properties:
```
Comment (mazhelez) [inline RELEASENOTES.md:9]:
```suggestion
- `trustedNuGetFeeds` - can be an array of NuGet feed specifications, which AL-Go for GitHub will use for dependency resolution. Every feed specification must include a URL property and can optionally include a few other properties:
```

File: RELEASENOTES.md [hunk 2]
```diff
@@ -2,6 +2,20 @@
...
+  - Url - The URL of the feed (examples: https://pkgs.dev.azure.com/myorg/apps/\_packaging/myrepo/nuget/v3/index.json or https://nuget.pkg.github.com/mygithuborg/index.json")
+  - Patterns - AL-Go for Github will only trust packages, where the ID matches this pattern. Default is all packages (\*).
+  - FingerPrints - If specified, AL-Go for GitHub will only trust packages signed with one of the fingerprints specified in this array of fingerprints.
```
Comment (mazhelez) [inline RELEASENOTES.md:12]: Wording could be better here: packages aren't _signed_ with fingerprints.

File: RELEASENOTES.md [hunk 3]
```diff
@@ -2,6 +2,20 @@
...
+  - AuthTokenSecret - If the NuGet feed specified by URL is private, the authTokenSecret must be the name of a secret containing the authentication token with permissions to search and read packages from the NuGet feed
+
+### Support for delivering to GitHub Packages and NuGet
+
+With this release the implementation for delivering to NuGet packages (by adding the NuGetContext secret), is similar to the functionality behind delivering to GitHub packages and the implementation is final.
```
Comment (mazhelez) [inline RELEASENOTES.md:17]: What does "the implementation is final" mean here?

File: Scenarios/secrets.md [hunk 4]
```diff
@@ -181,6 +181,12 @@ Example: `{"storageAccountName":"MyStorageName","storageAccountKey":"JHFZErCyfQ8
...
+If you create a secret called GitHubPackagesContext, then AL-Go for GitHub will automagically deliver apps to this NuGet feed after every successful build. AL-Go for GitHub will also use this NuGet feed for dependency resolution when building apps, giving you automatic dependency resolution within all your repositories sharing this secret.
 
 Example: `{"token":"ghp_NDdI2ExxxxxxxxxxxxxxxxxAYQh","serverUrl":"https://nuget.pkg.github.com/mygithuborg/index.json"}`
+
+## <a id="NuGetContext"></a>**NuGetContext** -> Deliver to NuGet
+
+If you create a secret called NuGetContext, then AL-Go for GitHub will automagically deliver apps to this NuGet feed after every successful build. AL-Go for GitHub will NOT use this NuGet feed for dependency resolution when building apps. If you want to use this feed for dependency resolution as well, you need to add this to the [trustedNuGetFeeds](https://aka.ms/algosettings#trustedNuGetFeeds) setting.
+
+Example: `{"token":"ydtaxlw66xxxxxxxxxxxxxxgqdq","serverUrl":"https://pkgs.dev.azure.com/myorg/apps/_packaging/myrepo/nuget/v3/index.json"}`
```
Comment (mazhelez) [inline Scenarios/secrets.md:192]:
```suggestion
Example: `{"token":"<token>","serverUrl":"https://pkgs.dev.azure.com/myorg/apps/_packaging/myrepo/nuget/v3/index.json"}`

```

Always avoid strings that resembles tokens, passwords, etc.. Even in docs and tests. It's so painful to deal with these if the scanning tool detects them.

File: Actions/RunPipeline/RunPipeline.ps1 [hunk 5]
```diff
@@ -113,6 +113,38 @@ try {
     $settings = AnalyzeRepo -settings $settings -baseFolder $baseFolder -project $project @analyzeRepoParams
     $settings = CheckAppDependencyProbingPaths -settings $settings -token $token -baseFolder $baseFolder -project $project
 
+    if ($bcContainerHelperConfig.ContainsKey('TrustedNuGetFeeds')) {
```
Comment (mazhelez) [inline Actions/RunPipeline/RunPipeline.ps1:116]: Could this be after the "Repository is empty, exiting" check?

**Review by mazhelez — COMMENTED** [review]
Looks good.

Left some comments, so one more iteration might be needed.

File: Actions/Deliver/Deliver.ps1 [hunk 6]
```diff
@@ -226,9 +225,22 @@ foreach ($thisProject in $projectList) {
...
     }
-    elseif ($deliveryTarget -eq "GitHubPackages") {
-        $githubPackagesCredential = [System.Text.Encoding]::UTF8.GetString([System.Convert]::FromBase64String($secrets.githubPackagesContext)) | ConvertFrom-Json
-        'Apps' | ForEach-Object {
+    elseif ($deliveryTarget -eq 'GitHubPackages' -or $deliveryTarget -eq 'NuGet') {
+        $preReleaseTag = ''
+        try {
+            $nuGetAccount = [System.Text.Encoding]::UTF8.GetString([System.Convert]::FromBase64String($secrets."$($deliveryTarget)Context")) | ConvertFrom-Json | ConvertTo-HashTable
+            if ($deliveryTarget -eq 'NuGet' -and $type -eq 'CD') {
+                $preReleaseTag = 'preview'
```
[the remark this answers is not in the document]
Comment (freddydk) [reply Actions/Deliver/Deliver.ps1:233]: added comments

File: Actions/RunPipeline/RunPipeline.ps1 (hunk 5 above)
Comment (freddydk) [reply Actions/RunPipeline/RunPipeline.ps1:116]: Yes

File: RELEASENOTES.md (hunk 2 above)
Comment (freddydk) [reply RELEASENOTES.md:12]: I actually think it is called thumbprint all over - renaming before we ship.
Also requires a new BcContainerHelper in a sec.

File: RELEASENOTES.md (hunk 3 above)
Comment (freddydk) [reply RELEASENOTES.md:17]: There was a POC implementation in version 2.1 which has been running since.

File: RELEASENOTES.md (hunk 2 above)
Comment (freddydk) [reply RELEASENOTES.md]: Corrected the dotnet tool for checking this uses fingerprints - I will stick with this implementation

**Review by mazhelez — APPROVED** [review]

**Review by aholstrup1 — APPROVED** [review]

**Pull request merged**